contracts/contracts/opl/Endpoint.sol

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+/// Unable to automatically configure OPL. Please use the manual version of the base contract.
+error AutoConfigUnavailable();
+/// The method cannot be called in this context.
+error Forbidden(bytes32 location);
+/// This message arrived too early or late.
+error WrongSeqNum(uint256 expected, uint256 got);
+/// The function was called with an invalid argument.
+error InvalidArgument(bytes32 location);
+
+/// The outcome of the message call.
+enum Result {
+    // The message was rejected.
+    PermanentFailure,
+    // The message was rejected but may be accepted later.
+    TransientFailure,
+    // The message was accepted and processed.
+    Success
+}
+
+interface ICelerMessageBus {
+    function feeBase() external view returns (uint256);
+
+    function feePerByte() external view returns (uint256);
+
+    function sendMessage(
+        address _host,
+        uint256 _hostChainId,
+        bytes calldata _message
+    ) external payable;
+}
+
+contract BaseEndpoint {
+    address private immutable bus;
+    bool private immutable inOrder;
+
+    address private host;
+    uint256 private hostChainId;
+
+    mapping(bytes32 => function(bytes calldata) returns (Result))
+        private endpoints;
+
+    uint256 private txSeq;
+    uint256 private rxSeq;
+
+    constructor(
+        address _host,
+        uint256 _hostChainId,
+        address _bus,
+        bool _inOrder
+    ) {
+        if (_host == address(0)) revert InvalidArgument("endpoint ctor");
+        if (_hostChainId == block.chainid || _hostChainId == 0)
+            revert InvalidArgument("endpoint ctor");
+        host = _host;
+        hostChainId = _hostChainId;
+        bus = _bus;
+        inOrder = _inOrder;
+    }
+
+    function registerEndpoint(
+        bytes memory _method,
+        function(bytes calldata) returns (Result) _cb
+    ) internal {
+        // This is a waste of an SLOAD, but the alternative before immutable arrays
+        // (https://github.com/ethereum/solidity/issues/12587) land is terribly verbose.
+        // This can be fixed once gas usage becomes a problem.
+        endpoints[bytes4(keccak256(_method))] = _cb;
+    }
+
+    function postMessage(bytes memory _method, bytes memory _message) internal {
+        uint256 fee = estimateFee(_message);
+        ICelerMessageBus(bus).sendMessage{value: fee}(
+            host,
+            hostChainId,
+            abi.encodeWithSelector(bytes4(keccak256(_method)), txSeq, _message)
+        );
+        ++txSeq;
+    }
+
+    function executeMessage(
+        address _sender,
+        uint64 _senderChainId,
+        bytes calldata _message,
+        address
+    ) external payable returns (uint256) {
```

Comment:
I assume we are avoiding Celer imports (e.g. for types, auth check modifiers) for a reason (overhead, flexibility)?

Reply:
`GPL-3.0-or-later` :P